### PR TITLE
FIX: Scrolling in one insert tab affects other tabs

### LINKED ISSUE #1926: Scrolling in one insert tab affects other tabs
version: 0.5

## Issue Overview
If I scroll in one of the insert tabs, it causes the content to be scrolled in the other tabs. See this gif: 

gutenburg-scroll mov

## Steps to Reproduce (for bugs)

1. Click Insert
2.Select Blocks
3. Scroll down
4. Select Embeds

## Expected Behavior

Each tab should scroll independently

## Current Behavior

Scrolling in one tab affects the other tabs

### PR DESCRIPTION
**Pull request by tg-ephox**
While the Inserter remains open, the scroll position is retained for separate tabs if they're scrolled then switched. Once it's closed and re-opened the scroll position will start at top again.

fixes: https://github.com/WordPress/gutenberg/issues/1926

### CODE REVIEW

File: editor/inserter/menu.js [hunk 1]
```diff
@@ -330,7 +341,7 @@ export class InserterMenu extends Component {
 					ref={ this.bindReferenceNode( 'search' ) }
 					tabIndex="-1"
 				/>
-				<div role="menu" className="editor-inserter__content">
+				<div ref={ ( r ) => this._tabContainer = r } role="menu" className="editor-inserter__content">
```
Comment (youknowriad) [inline editor/inserter/menu.js]: Do you think we can remove the `_`, we're not using this convention for the private properties?
Can we avoid small variables like `r` maybe `ref` ?

**Review by youknowriad — COMMENTED** [review]
Thanks for fixing this, it works great.
Do you think we could extract this to a `HoC` or an utility to apply it elsewhere if needed?

File: editor/inserter/menu.js (hunk 1 above)
Comment (aduth) [inline editor/inserter/menu.js]: Would also be good to break up into multiple lines.

File: editor/inserter/menu.js [hunk 2]
```diff
@@ -305,6 +313,9 @@ export class InserterMenu extends Component {
 	}
 
 	switchTab( tab ) {
+		// store the scrollTop of the tab switched from
+		this.tabScrollTop[ this.state.tab ] = this._tabContainer.scrollTop;
+		this.didSwitchTab = true;
```
Comment (aduth) [inline editor/inserter/menu.js]: I don't think we need this variable. Instead, `componentDidUpdate` receives the previous state as an argument, so we can check whether tab changed:

```js
componentDidUpdate( prevProps, prevState ) {
	// ...

	const didSwitchTab = this.state.tab !== prevState.tab;
	if ( didSwitchTab ) {
		// ...
```

**Review by aduth — COMMENTED** [review]
An improvement on the current situation, but still not entirely convinced we need per-tab-scroll-memory https://github.com/WordPress/gutenberg/issues/1926#issuecomment-316177784

**Comment by aduth** [comment]
>Do you think we could extract this to a HoC or an utility to apply it elsewhere if needed?

Not entirely sure this is warranted. Not sure the general use case has presented itself.

File: editor/inserter/menu.js (hunk 2 above)
Comment (tg-ephox) [inline editor/inserter/menu.js]: You're right, that's a lot cleaner!

**Comment by tg-ephox** [comment]
@aduth I was thinking of this from the perspective of browser tabs, where the scroll position is obviously maintained when tabs are switched. However I'm happy to change this to reset `scrollTop` when tabs are switched if this is preferred.  I also made the other changes you noted in the review.

File: editor/inserter/menu.js [hunk 3]
```diff
@@ -47,6 +47,9 @@ export class InserterMenu extends Component {
 		this.getBlocksForCurrentTab = this.getBlocksForCurrentTab.bind( this );
 		this.sortBlocks = this.sortBlocks.bind( this );
 		this.addRecentBlocks = this.addRecentBlocks.bind( this );
+
+		this.tabScrollTop = { recent: 0, blocks: 0, embeds: 0 };
+		this.didSwitchTab = false;
```
Comment (aduth) [inline editor/inserter/menu.js]: This is unused now and should be removed.

**Review by aduth — CHANGES_REQUESTED** [review]
This works really well. Aside from one unused variable, this looks good to merge.

**Review by aduth — APPROVED** [review]

**Pull request merged**